### PR TITLE
Added test for pendulum demo

### PR DESCRIPTION
**Pull request by esteve**
Depends on ros2/launch#18

Connects to #43

### CODE REVIEW

**Comment by esteve** [comment]
@jacquelinekay I couldn't get this to work the dynamic variant of connext, should it work?

**Comment by jacquelinekay** [comment]
I pushed a rebase of this since it wasn't compiling. Testing for Connext dynamic now.

File: pendulum_control/CMakeLists.txt [hunk 1]
```diff
@@ -66,6 +66,48 @@ install(
...
+    find_package("${middleware_impl}" REQUIRED)
+  endforeach()
+
+  set(RCLCPP_DEMO_PENDULUM_LOGGER_EXPECTED_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/test/pendulum_logger")
+  set(RCLCPP_DEMO_PENDULUM_DEMO_EXPECTED_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/test/pendulum_demo")
+  set(RCLCPP_DEMO_PENDULUM_TELEOP_EXPECTED_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/test/pendulum_teleop")
+
+  foreach(middleware_impl ${middleware_implementations})
+    # get typesupport of rmw implementation to include / link against the corresponding interfaces
+    get_rmw_typesupport(typesupport_impl "${middleware_impl}")
```
Comment (dirk-thomas) [inline pendulum_control/CMakeLists.txt]: This variable seems to be unused?

File: pendulum_control/test/test_pendulum_demo.py.in [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        output_handlers=[pendulum_teleop_handler],
+    )
+
+    launcher = DefaultLauncher()
+    launcher.add_launch_descriptor(launch_descriptor)
+    rc = launcher.launch()
+
+    assert rc == 0, \
+        "The launch file failed with exit code '" + str(rc) + "'. " \
+        'Maybe the client did not receive any messages?'
```
Comment (dirk-thomas) [inline pendulum_control/test/test_pendulum_demo.py.in]: Does the second part of the message make any sense here?

**Comment by jacquelinekay** [comment]
@esteve, I ran the executable manually and the connext_dynamic_cpp version works as expected, but the test times out. I'll investigate now.

Also, should I ever expect to be able to run launch manually? The only time it seems to work is when I invoke it through `ament test`. If I try to run `launch test_<testname>.py`, the launch executable fails to find the right module.

File: pendulum_control/test/test_pendulum_demo.py.in [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    )
+
+    pendulum_demo_executable = '@RCLCPP_DEMO_PENDULUM_DEMO_EXECUTABLE@'
+    pendulum_demo_output_file = '@RCLCPP_DEMO_PENDULUM_DEMO_EXPECTED_OUTPUT@'
+    pendulum_demo_name = 'test_executable_1'
+    pendulum_demo_handler = create_handler(pendulum_demo_name, launch_descriptor, pendulum_demo_output_file)
+    assert pendulum_demo_handler, 'Cannot find appropriate handler for %s' % pendulum_demo_output_file
+    output_handlers.append(pendulum_demo_handler)
+    launch_descriptor.add_process(
+        cmd=[pendulum_demo_executable, '-i', '0'],
```
Comment (jacquelinekay) [inline pendulum_control/test/test_pendulum_demo.py.in]: I'm not sure how launch works, so bear with me--specifying `-i 0` will make the executable run indefinitely until interrupted. When will the pendulum_demo executable get interrupted?

**Comment by dirk-thomas** [comment]
Are your sourcing the setup file before trying to invoke the launch file directly?

**Comment by jacquelinekay** [comment]
Yes.

**Comment by dirk-thomas** [comment]
The command line tool `launch` only works for launch files (which need to have a `launch` function). But these tests don't have a function named `launch` but potentially multiple `test_*` functions.

**Comment by jacquelinekay** [comment]
@esteve, I think the test failures we both saw for Connext Dynamic have to do with nondeterministic startup behaviors.

In your configuration, you set the pendulum_demo test to go forever and the matching condition to be the statement that is printed when the subscription callback for the topic published by `pendulum_teleop` is received.

Note that the pendulum_teleop publishes a message once and exits (not a great design, if we had real-time services I would use that instead).

However I think that in the Connext Dynamic implementation, the pendulum_teleop process initializes first and sends the message before the pendulum_demo process is even finished initializing. Therefore the demo process never receives the message and the test times out.

With that in mind, I've split the test into two separate nose tests:

1. run the pendulum_demo for a limited number of iterations (1000, should be at least 1 second) and the pendulum_logger. The output regex matches the format that pendulum_demo should produce from start to finish and the exact output made by pendulum_logger in 1 iteration. Therefore the process should terminate when pendulum_demo is finished executing.

2. run the pendulum_demo for an unlimited number of iterations (-i 0) and run the pendulum_teleop program. Matches the printout made in the subscription callback for pendulum_teleop (that is, the same output that you originally proposed in this PR). I would expect this test to still be flaky for the same reason I stated above, but from my local testing it passes reliably.

Please test this locally and merge if you think it looks good:

https://github.com/ros2/demos/compare/pendulum_demo_test...pendulum_demo_test_jackie

**Comment by dirk-thomas** [comment]
Since `launch_testing` only implements partial matching yet (https://github.com/ros2/launch/pull/13#discussion_r44998991) I don't think you multiple lines with regex won't work as expected.

**Comment by jacquelinekay** [comment]
oh, whoops. I guess I could match only the last line of the expected output...

How hard would it be to fully implement matching?

File: pendulum_control/CMakeLists.txt (hunk 1 above)
Comment (esteve) [inline pendulum_control/CMakeLists.txt]: Removed.

File: pendulum_control/test/test_pendulum_demo.py.in (hunk 2 above)
Comment (esteve) [inline pendulum_control/test/test_pendulum_demo.py.in]: Removed.

**Comment by esteve** [comment]
@jacquelinekay thanks for looking into this! I addressed @dirk-thomas 's feedback and pushed a few more commits to this branch. Shall I merge your changes or you'd prefer to rebase on top of this branch?

**Comment by jacquelinekay** [comment]
Whatever you think is best. The branch in its current state depends on
ros2/launch#19 though. I am working on a test for that PR right now, but
I'm a bit unfamiliar with how everything works--I tried to add a coroutine
to the launch descriptor, but the InMemoryHandler doesn't seem to be
processing anything the coroutine prints to stdout. Do you have any advice
here? Should I just write a separate process and launch that in the test?

On Tue, Dec 15, 2015 at 10:25 AM, Esteve Fernandez <notifications@github.com

File: pendulum_control/CMakeLists.txt [hunk 4]
```diff
@@ -66,6 +66,45 @@ install(
...
+      test_pendulum__${middleware_impl}.py
+      @ONLY
+    )
+
+    ament_add_nose_test(test_demo_pendulum__${middleware_impl}
+      "${CMAKE_CURRENT_BINARY_DIR}/test_pendulum__${middleware_impl}.py"
+      TIMEOUT 30)
+    set_tests_properties(test_demo_pendulum__${middleware_impl}
+    PROPERTIES DEPENDS "test_demo_pendulum__${middleware_impl} test_demo_pendulum__${middleware_impl}")
+  endforeach()
```
Comment (dirk-thomas) [inline pendulum_control/CMakeLists.txt]: You might want to use the new CMake function `call_for_each_rmw_implementation` for this (https://github.com/ros2/rmw/pull/50).

**Comment by jacquelinekay** [comment]
@esteve I rebased my branch, you should be good to merge it in.

I think that we should then run CI and I will +1 if CI looks good.

**Comment by esteve** [comment]
@jacquelinekay I've merged your branch into this, I'll take it from here. Thanks!

File: pendulum_control/CMakeLists.txt (hunk 1 above)
Comment (dirk-thomas) [inline pendulum_control/CMakeLists.txt]: This line is still in the current patch.

Comment (esteve) [inline pendulum_control/CMakeLists.txt]: Removed again.

**Comment by esteve** [comment]
This passes locally on Linux, I'll trigger CI jobs for OSX and Windows.

**Comment by esteve** [comment]
CI jobs:

http://ci.ros2.org/job/ci_linux/745/
http://ci.ros2.org/job/ci_osx/632/
http://ci.ros2.org/job/ci_windows/816/

**Comment by dirk-thomas** [comment]
I can't find the tests on OS X or Windows anywhere?
* Linux: http://ci.ros2.org/job/ci_linux/745/testReport/(root)/ test_pendulum_*
* OS X: http://ci.ros2.org/job/ci_osx/632/consoleFull#console-section-158
* Windows: http://ci.ros2.org/job/ci_windows/816/consoleFull#console-section-158 http://ci.ros2.org/job/ci_windows/ws/ws/build/pendulum_control/Testing/20151216-2303/Test.xml/*view*/

**Comment by jacquelinekay** [comment]
Building this demo is disabled on OSX and Windows.

**Comment by dirk-thomas** [comment]
@jacquelinekay Ok, since this patch contains Windows specific code and it had Windows and OS X jobs I didn't think about it.

File: pendulum_control/CMakeLists.txt [hunk 5]
```diff
@@ -66,6 +66,59 @@ install(
 if(AMENT_ENABLE_TESTING)
   find_package(ament_lint_auto REQUIRED)
   ament_lint_auto_find_test_dependencies()
+
+  if(NOT WIN32)
```
Comment (jacquelinekay) [inline pendulum_control/CMakeLists.txt]: This if statement is unnecessary because the Cmake file already exits if WIN32 at the top of the file.

**Comment by jacquelinekay** [comment]
Besides my small comment about the CMakeLists, +1.

**Comment by esteve** [comment]
@jacquelinekay I've removed that logic and wrote a comment detailing the reasoning for those variables.

**Comment by esteve** [comment]
Last CI job:

http://ci.ros2.org/job/ci_linux/747/

**Comment by esteve** [comment]
The demo tests pass, yay!

**Pull request merged**